arangod/Aql/IResearchViewExecutor.cpp

```diff
@@ -0,0 +1,591 @@
+           "arangosearch view, cid '"
+        << cid << "'";
+    query.registerWarning(TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND, msg.str());
+    return false;
+  }
+
+  // read document from underlying storage engine, if we got an id
+  if (collection->readDocumentWithCallback(query.trx(), documentId, ctx.callback)) {
+    // in the ordered case we have to write scores as well as a document
+    if /* constexpr */ (ordered) {
+      // scorer register are placed consecutively after the document output register
+      RegisterId scoreReg = ctx.docOutReg + 1;
+
+      for (auto& it : _indexReadBuffer.getScores(bufferEntry)) {
+        TRI_ASSERT(infos().isScoreReg(scoreReg));
+        bool mustDestroy = false;
+        AqlValueGuard guard{it, mustDestroy};
+        ctx.outputRow.moveValueInto(scoreReg, ctx.inputRow, guard);
+        ++scoreReg;
+      }
+
+      // we should have written exactly all score registers by now
+      TRI_ASSERT(!infos().isScoreReg(scoreReg));
+    }
+
+    return true;
+  }
+
+  return false;
+}
+
+template <bool ordered>
+void IResearchViewExecutor<ordered>::evaluateScores(ReadContext& ctx) {
+  // This must not be called in the unordered case.
+  TRI_ASSERT(ordered);
+
+  // evaluate scores
+  _scr->evaluate();
+
+  // in arangodb we assume all scorers return float_t
+  auto begin = reinterpret_cast<const float_t*>(_scrVal.begin());
+  auto end = reinterpret_cast<const float_t*>(_scrVal.end());
+
+  // scorer register are placed consecutively after the document output register
+  // is used here currently only for assertions.
+  RegisterId scoreReg = ctx.docOutReg + 1;
+
+  // copy scores, registerId's are sequential
+  for (; begin != end; ++begin, ++scoreReg) {
+    TRI_ASSERT(infos().isScoreReg(scoreReg));
+    _indexReadBuffer.pushScore(*begin);
+  }
+
+  // We should either have no _scrVals to evaluate, or all
+  TRI_ASSERT(begin == nullptr || !infos().isScoreReg(scoreReg));
+
+  while (infos().isScoreReg(scoreReg)) {
+    _indexReadBuffer.pushScoreNone();
+    ++scoreReg;
+  }
+
+  // we should have written exactly all score registers by now
+  TRI_ASSERT(!infos().isScoreReg(scoreReg));
+  TRI_ASSERT(scoreReg - ctx.docOutReg - 1 == infos().getNumScoreRegisters());
+}
+
+template <bool ordered>
+void IResearchViewExecutor<ordered>::fillBuffer(IResearchViewExecutor::ReadContext& ctx) {
+  TRI_ASSERT(_filter != nullptr);
+
+  std::size_t const atMost = ctx.outputRow.numRowsLeft();
+
+  size_t const count = _reader->size();
+  for (; _readerOffset < count;) {
+    if (!_itr) {
+      if (!_indexReadBuffer.empty()) {
+        // We may not reset the iterator and continue with the next reader if we
+        // still have documents in the buffer, as the cid changes with each
+        // reader.
+        break;
+      }
+
+      if (!resetIterator()) {
+        continue;
+      }
+
+      // CID is constant until the next resetIterator()
+      _indexReadBuffer.setCidAndReset(_reader->cid(_readerOffset));
```

Comment:
we have to lookup collection by cid and store it in `_indexReadBuffer` for every segment rather than for every document:
```
Query& query = infos().getQuery();
auto collection = lookupCollection(*query.trx(), _reader->cid(_readerOffset), query);
if (!collection) {
  // log
  continue;
}

_indexReadBuffer.setCidAndReset(collection.get());
```